numba/tests/test_slices.py

```diff
@@ -78,6 +83,25 @@ def test_slice_constructor(self):
             got = cfunc(*args)
             self.assertPreciseEqual(got, expected)
 
+    def test_slice_indices(self):
+        """Test that a numba slice returns same result for .indices as a python one."""
+        N_TESTS = 500
+
+        lengths = np.random.randint(0, 100, N_TESTS)
```

Comment:
I am not sure that random fuzzing is the correct approach. The problem with randomness is that it may be hard to repeat a failed test if the randomness manages to fuzz out an exotic and highly improbable combination of arguments. A well defined sequence of consistent input might be the wiser choice here.

Reply:
Agreed. I've replaced it with more comprehensive coverage on a smaller input space. This ends up being a larger set of test cases, and could probably be reduced a bit if runtime is an issue (it's ~0.25 seconds for `python -m numba.runtests -k "test_slice_indices"`). I've also added assertions to make sure the right exceptions were being thrown, and it turns out they weren't.

Reply:
Maybe, the running time of our test-suite is an issue for sure. There is an open PR here that adds instrumentation to the test-suite:

https://github.com/numba/numba/pull/4640

I would tend to argue that 0.25 seconds is acceptable if it is a good balance. There are other test cases that are much worse offenders in terms of runtime.

Reply:
> I've also added assertions to make sure the right exceptions were being thrown, and it turns out they weren't.

What we usually do, is have one set of known good inputs that we use to test the so-called 'happy-path'. Which is when the function does what it is supposed to. And another set of
tests that execute the function with input designed to raise exceptions. Usually, we check those using a `with self.asserttRaises` clause. Also, raising exceptions creates reference leaks, so you will need to call `self.disable_leak_check()` before checking those. There are numerous examples of this in the Numba test-code so it sholdn't be to hard to find something to crib off. For example this one:

https://github.com/numba/numba/blob/master/numba/tests/test_np_functions.py#L3067

Reply:
I could definitely add some explicit test cases here, and the `self.disable_leak_check()`.

From a fairly painful experience of trying to replicate slicing behaviour, I think it's hard to fully understand all possible cases. I'd like to keep the 'range of inputs against an oracle' test as it gives me much more confidence I haven't missed something.

Reply:
Do I need `disable_leak_check` here, as it doesn't currently subclass from `MemoryLeak`?

Reply:
Yes, you need to subclass from `MemoryLeakMixin` to enable leak-checking.

Reply:
I've added the mixin and disabled checks on the tests which use exceptions.